features/prog.go

```diff
@@ -0,0 +1,157 @@
+package features
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/asm"
+	"github.com/cilium/ebpf/internal"
+	"github.com/cilium/ebpf/internal/unix"
+)
+
+func init() {
+	pc.progTypes = make(map[ebpf.ProgramType]error)
+}
+
+var (
+	pc progCache
+)
+
+type progCache struct {
+	sync.Mutex
+	progTypes map[ebpf.ProgramType]error
+}
+
+func createProgLoadAttr(pt ebpf.ProgramType) (*internal.BPFProgLoadAttr, error) {
+	var expectedAttachType ebpf.AttachType
+
+	insns := asm.Instructions{
```

Comment:
The assembly of the probe program could be moved to `init()` (rather, a function called by `init()`), since it only needs to be assembled once. `instructions` and the parameter to `InsCount` could then be stored in `pc`, which saves some garbage.

Reply:
Since I probably also want to use that function to generate the load attribute when probing helpers, I imagined generating the instructions "dynamically" here according to the helper call thats supposed to be probed. So moving this to `init()` would just be reverted in that case.

Reply:
I was thinking we could use the pre-cooked program when probing program types and generating one on-the-fly when probing helper types. This might be a premature optimization, though. Let's wait and see what this will look like with helper probes added. :+1: